### PR TITLE
Add support for composable index templates

### PR DESCRIPTION
**Pull request by NITEMAN**
#### Pull Request (PR) description
Add support for composable index templates

### CODE REVIEW

**Comment by NITEMAN** [comment]
I don't know why "Puppet 6 - Ubuntu 18.04" acceptance is failing but AFAIK it seems related with elastic_stack module and this PR is working as expected

File: manifests/init.pp
Comment (bastelfreak) [inline manifests/init.pp:478]: instead of `create_resources()`, could you iterate on the hash like:
```puppet
$elasticsearch::component_templates.each |String $key, Hash $values| {
  elasticsearch::component_template { $key:
    * => $values,
}
```

that pattern is now preferred over `create_resources()`

File: README.md
Comment (bastelfreak) [inline README.md:268]: can you maybe extend the description a bit and mention the new types/defined resources? maybe also mention which elasticsearch versions are supported?

**Comment by bastelfreak** [comment]
hi @NITEMAN, 
thanks for the awesome work. The type/provider, puppet and test code look all pretty solid. I made some tiny inline comments but overall this looke quite good.

File: README.md
Comment (NITEMAN) [reply README.md]: I've expanded it a bit, let me know if there is still something missing/unclear

**Comment by NITEMAN** [comment]
Hi @bastelfreak , thanks to all of you for taking over with this module!

Most code is copied/adapted from other parts of the module so I hope my bits are ok... for me the hard part has come dealing with testing so please double check these parts

File: data/common.yaml
Comment (bastelfreak) [inline data/common.yaml:12]: ah another tiny thing. could you please not add new datat to the common.yaml but directly to the init.pp? That enables puppet-string to pickup the default value. That's our new best practice at vox pupuli, we just had no time yet to migrate all common.yaml files to init.pp

File: README.md
Comment (bastelfreak) [reply README.md]: thanks, looks good now.

File: data/common.yaml
Comment (NITEMAN) [reply data/common.yaml]: Done, I thought PDK complains about this pattern, but for me is ok ;)

**Comment by bastelfreak** [comment]
thanks! looks good to me now. I added a few reviewers that have more knowledge about elasticsearch/ruby.

**Review by crazymind1337 — COMMENTED** [review]
The puppet code looks fine for me. Since I am not into writing provider/types via rb files I cannot make any reliable statement about that.

**Review by smortex — APPROVED** [review]
CI failure is unrelated. Loads of code with aspects of ES I do not know, but the tests seems to prove it is okay :smile:

**Comment by NITEMAN** [comment]
Thanks, I'm now confident enough to try to reroll #1103 and #1103 adding tests

**Comment by NITEMAN** [comment]
Removed some leftover todos, add more tests, and fixed some comments.

Any chances for this to be merged soon?

**Comment by NITEMAN** [comment]
Test fail seems unrelated/accidental, but I've no idea of how to re-run test without force-pushing

**Comment by usp-rol** [comment]
@smortex  since you are a member having write access, all tests seem fine and we would be excited to see #1163, #1164 and this one here make it into the master...please could you help find the missing peace (or help out whats missing?)

thank you very much!

**Comment by NITEMAN** [comment]
Hi @smortex 

My original plan was to rebase #1163 after #1146 merge... and then doing the same with #1164 , that would fit git / github keeping  changes atomic... but I'll try your suggestion tomorrow (never done before something like that)

Best regards!

**Comment by usp-rol** [comment]
Thank you both @smortex  and @NITEMAN  so much! We're happy to this see these features in a new release.

**Comment by NITEMAN** [comment]
@smortex I haven't managed to gess how to "foce-push x into y", what I've done is

- Rebase new-templates (#1146) into current master (`git rebase -i upstream/master`)
- Rebase ilm_policies (#1163) into new-templates (`git rebase -i new-templates`)
- Rebase slm_policies (#1164) into ilm_policies (`git rebase -i ilm_policies`)

That should fix any commit reference issues

But now I'm again stuck on an unrelated CI fail (for me it's quite frustrating), I suppose master itself fails "CI / Puppet / Static validations"

**Comment by bastelfreak** [comment]
hi @NITEMAN , you can regenerate the REFERENCE.md with `bundle exec rake reference`

**Comment by NITEMAN** [comment]
Many thanks @bastelfreak , both for the comment and for your support on IRC... I hope everything will came together now

**Comment by NITEMAN** [comment]
@smortex , @bastelfreak  Everything should be fine right now (all branches rebased updating reference on each commit)

Please let me know if there is something missing/needed

**Pull request merged**
